parcels/collection/collectionaos.py

```diff
@@ -484,6 +533,206 @@ def add_same(self, same_class):
+                min_pu = mpi_comm.bcast(min_pu, root=0)
+                self._pu_indicators = mpi_comm.bcast(self._pu_indicators, root=0)
+                pu_indices = np.nonzero(min_pu == mpi_rank)[0]
+                ids = np.arange(ScipyParticle.lastID, stop=ScipyParticle.lastID+data_array['lon'].shape[0]) if 'id' not in data_array.keys() else data_array['id']
+                mpi_comm.Bcast(ids, root=0)
+                pu_ids = ids
+                new_lastID = 0
+                if mpi_rank == 0:
+                    new_lastID = ScipyParticle.lastID+data_array['lon'].shape[0]-1
+                new_lastID = mpi_comm.bcast(new_lastID, root=0)
+                self._pclass.setLastID(new_lastID)
+                pu_center = np.array(np.mean(spdata, axis=0), dtype=self._lonlatdepth_dtype)
+                n_pu_data = pu_indices.shape[0]
+                pu_ncenters = None
+                if mpi_rank == 0:
+                    pu_ncenters = np.empty([mpi_size, pu_center.shape[0]], dtype=self._latlondepth_dtype)
+                mpi_comm.Gather(pu_center, pu_ncenters, root=0)
+                pu_ndata = mpi_comm.gather(n_pu_data, root=0)
+                if mpi_rank == 0:
+                    for i in range(self._pu_centers.shape[0]):
+                        ax = float(pu_ndata[i]) / float(len(np.nonzero(self._pu_indicators == i)[0]))
+                        self._pu_centers[i, :] += ax*pu_ncenters[i, :]
+                mpi_comm.Bcast(self._pu_centers, root=0)
+            else:
+                pu_ids = np.arange(ScipyParticle.lastID, stop=ScipyParticle.lastID+data_array['lon'].shape[0]) if 'id' not in data_array.keys() else data_array['id']
+                new_lastID = ScipyParticle.lastID+data_array['lon'].shape[0]-1
+                self._pclass.setLastID(new_lastID)
+                pu_indices = np.arange(start=0, stop=data_array['lon'].shape[0])
+                n_pu_data = pu_indices.shape[0]
+            if n_pu_data <= 0:
+                results = []
+            else:
+                for i in pu_indices:
+                    pid = pu_ids[i]
+                    pdata = self._pclass(lon=data_array['lon'][i], lat=data_array['lat'][i], pid=pid, ngrids=self._ngrid)
+                    if 'depth' in data_array.keys():
+                        pdata.depth = data_array['depth'][i]
+                    if 'time' in data_array.keys():
+                        pdata.time = data_array['time'][i]
+                    if 'dt' in data_array.keys():
+                        pdata.dt = data_array['dt'][i]
+                    for key in self._kwarg_keys:
+                        if key in data_array.keys():
+                            setattr(pdata, key, data_array[key][i])
+                    results.append(self.add_single(pdata, pu_checked=True))
+        self._ncount = len(self._data)
+        return results
+
+    def add_single(self, particle_obj, pu_checked=False):
```

Comment:
A lot of the code in this function seems similar to that of `add_multiple()` above. Is there a way to reuse some of the code, to simplify its maintenance?

Reply:
it looks similar, but in terms of performance, fusing the two is disadvantageous. Because the `add_multiple(...)` function is provided with several Numpy arrays, they can be more efficiently transmitted via message passing (i.e. `MPI`). Hence the code makes use of different `MPI` calls (-> efficiency of network-adapter transmission). If one wants to add a single particle, e.g. due to fragmentation, then it is possible to do so via `add_single`.

Initially, this function was declared as `NotImplementedError` in previous versions. It is mandatory for the working of node-based lists though - irrespective of the fragmentation use-case. As discussed, one requirement was to have all (currently) 3 collection types being able to perform the same functions, to avoid large explanation to users why this-and-that `add` function is available in the one collection but not in the other - especially if it's feasible to do for all 3 collections. It was one of the to-do items we agreed upon above. Still, as the objective with parcels and with the class hierarchy was to provide efficient methods for each individual collections, the current function design of `add_single(...)` and `add_multiple(...)`is the result of the specification.

TL;DR: sadly, this is the maximum degree of modularisation we can provide with the goal of performance and efficiency of both methods in mind. I hope it explains the design choice,

Reply:
OK, thanks for the explanation